q2_types/feature_data/_format.py

```diff
@@ -299,11 +300,60 @@ def validate(self, *args):
     'DifferentialDirectoryFormat', 'differentials.tsv', DifferentialFormat)
 
 
+class ProteinFASTAFormat(model.TextFileFormat):
+    def _validate_(self, level):
+        record_count_map = {'min': 5, 'max': None}
+        self._validate(record_count_map[level])
+
+    def _validate(self, n_records=None):
+        # read in using skbio and iterate over the contents -
+        # ValueErrors will be raised for wrong records
+        generator = self._read_protein_fasta(str(self))
+        if n_records is not None:
+            generator = itertools.islice(generator, n_records)
+        [x for x in generator]
+
+    def _read_protein_fasta(self, path):
+        return skbio.read(path, format='fasta', constructor=skbio.Protein)
+
+
+ProteinSequencesDirectoryFormat = model.SingleFileDirectoryFormat(
+    'ProteinSequencesDirectoryFormat',
+    'protein-sequences.fasta',
+    ProteinFASTAFormat)
+
+
+class AlignedProteinFASTAFormat(model.TextFileFormat):
+    def _validate_(self, level):
+        record_count_map = {'min': 5, 'max': None}
+        self._validate(record_count_map[level])
+
+    def _validate(self, n_records=None):
+        # read in using skbio and iterate over the contents -
+        # ValueErrors will be raised for wrong records
+        generator = self._read_protein_alignment_fasta(str(self))
+        if n_records is not None:
+            generator = itertools.islice(generator, n_records)
+        [x for x in generator]
+
+    def _read_protein_alignment_fasta(self, path):
+        return skbio.read(path, format='fasta',
+                          constructor=skbio.Protein, into=skbio.TabularMSA)
```

Comment:
Is this a greedy load? If so, that might be prohibitively expensive when the framework validates the format (which is pretty often).

Reply:
Good point - I missed that. Yes, when specifying `into` a generator wouldn't be returned anymore... If I don't specify that an "alignment" is required here it would still work in most cases (and would validate it correctly as an alignment), but when a pure protein FASTA file would be provided it would also parse it correctly (as it doesn't know whether protein or aligned protein is expected). I think for us that's an issue as we want to ensure that the user only provides an aligned protein sequences. I couldn't think of any other solution rather than rewriting that entirely in a way similar to AlignedDNAFASTAFormat. Do you think it would make sense to refactor `_validate_DNAFASTAFormats` into a more generic `_validate_FASTAFormats` method so that it can accept any FASTAValidator (depending on the alphabet required)? Then it shouldn't be that complicated to change that.

Reply:
> Do you think it would make sense to refactor _validate_DNAFASTAFormats into a more generic _validate_FASTAFormats method so that it can accept any FASTAValidator (depending on the alphabet required)?

Yeah, that sounds like a pretty reasonable approach. ccing @gregcaporaso - do I recall an open issue somewhere about another FASTA format need? I seem to recall a bunch of the validation logic was duplicated (for a deadline?) with the goal of merging back in, later. Probably a separate project, if that's the case, but would like to consider all the edges right now.

Reply:
Yep - good memory! 

In genome-sampler we have [an open issue](https://github.com/caporaso-lab/genome-sampler/issues/7) to reduce duplication of code related to fasta validation. We needed to support importing of fasta files with lowercase characters and non-IUPAC characters (both are common in GISAID). We [defined a new format](https://github.com/caporaso-lab/genome-sampler/blob/2020.8.0/genome_sampler/common.py#L72) there with a validate function that removes all of the checks on the characters contained in sequences. It'd be great to consolidate these, and maybe the current genome-sampler one could be used as the least restrictive validator. 

Let me know what I can do to help!

Reply:
OK, thanks for input @gregcaporaso and @thermokarst ! I can try to consolidate all those and re-write that a little bit. Would it maybe be better to have another PR in q2-types for that though and then, after the formats are cleaned up, come back here and change that accordingly?